<a href="https://colab.research.google.com/github/YLysov0017/Pproject/blob/master/ML_lab1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [1]:
import pandas as pd

In [7]:
df = pd.read_csv("/content/crimes_against_women_2001-2014.csv")
df

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
0               0  ANDHRA PRADESH           ADILABAD  2001    50   
1               1  ANDHRA PRADESH          ANANTAPUR  2001    23   
2               2  ANDHRA PRADESH           CHITTOOR  2001    27   
3               3  ANDHRA PRADESH           CUDDAPAH  2001    20   
4               4  ANDHRA PRADESH      EAST GODAVARI  2001    23   
...           ...             ...                ...   ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014     1   
10673         833     Lakshadweep  Total District(s)  2014     1   
10674         834      Puducherry           Karaikal  2014     3   
10675         835      Puducherry         Puducherry  2014     7   
10676         836      Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
0                            30            16   
1                            30             7   
2                            34            14   
3                            20            17   
4                            26            12   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
0                                                    149     
1                                                    118     
2                                                    112     
3                                                    126     
4                                                    109     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                              34                                  175   
1                              24                                  154   
2                              83                                  186   
3                              38                                   57   
4                              58                                  247   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[10677 rows x 11 columns]

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [8]:
#Колонки
df.columns

Index(['Unnamed: 0', 'STATE/UT', 'DISTRICT', 'Year', 'Rape',
       'Kidnapping and Abduction', 'Dowry Deaths',
       'Assault on women with intent to outrage her modesty',
       'Insult to modesty of Women', 'Cruelty by Husband or his Relatives',
       'Importation of Girls'],
      dtype='object')

In [9]:
#Первые 5 значений
df[:5]

Unnamed: 0        STATE/UT       DISTRICT  Year  Rape  \
0           0  ANDHRA PRADESH       ADILABAD  2001    50   
1           1  ANDHRA PRADESH      ANANTAPUR  2001    23   
2           2  ANDHRA PRADESH       CHITTOOR  2001    27   
3           3  ANDHRA PRADESH       CUDDAPAH  2001    20   
4           4  ANDHRA PRADESH  EAST GODAVARI  2001    23   

   Kidnapping and Abduction  Dowry Deaths  \
0                        30            16   
1                        30             7   
2                        34            14   
3                        20            17   
4                        26            12   

   Assault on women with intent to outrage her modesty  \
0                                                149     
1                                                118     
2                                                112     
3                                                126     
4                                                109     

   Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                          34                                  175   
1                          24                                  154   
2                          83                                  186   
3                          38                                   57   
4                          58                                  247   

   Importation of Girls  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0

In [10]:
#Последние 5 значений
df[len(df)-5:len(df)]

Unnamed: 0     STATE/UT           DISTRICT  Year  Rape  \
10672         832  Lakshadweep        Lakshadweep  2014     1   
10673         833  Lakshadweep  Total District(s)  2014     1   
10674         834   Puducherry           Karaikal  2014     3   
10675         835   Puducherry         Puducherry  2014     7   
10676         836   Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0

In [15]:
#Статистику

df.describe()

Unnamed: 0          Year          Rape  Kidnapping and Abduction  \
count  10677.000000  10677.000000  10677.000000              10677.000000   
mean    3871.570198   2007.697949     57.989885                 69.888358   
std     2816.369918      4.046874    214.230398                311.623450   
min        0.000000   2001.000000      0.000000                  0.000000   
25%     1009.000000   2004.000000      8.000000                  6.000000   
50%     3678.000000   2008.000000     22.000000                 20.000000   
75%     6347.000000   2011.000000     44.000000                 49.000000   
max     9016.000000   2014.000000   5076.000000              10626.000000   

       Dowry Deaths  Assault on women with intent to outrage her modesty  \
count  10677.000000                                       10677.000000     
mean      20.181699                                         113.539196     
std       98.276531                                         458.903951     
min        0.000000                                           0.000000     
25%        1.000000                                          10.000000     
50%        5.000000                                          34.000000     
75%       16.000000                                          85.000000     
max     2469.000000                                       10001.000000     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
count                10677.000000                         10677.000000   
mean                    27.419313                           209.224314   
std                    167.806797                           905.664362   
min                      0.000000                             0.000000   
25%                      0.000000                            11.000000   
50%                      2.000000                            50.000000   
75%                     12.000000                           144.000000   
max                   4970.000000                         23278.000000   

       Importation of Girls  
count          10677.000000  
mean               0.175330  
std                2.228637  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max               83.000000

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





1.   **Ответ 1.** 3871.570198	2007.697949	57.989885	69.888358	20.181699	113.539196	27.419313	209.224314	0.175330
2.   **Ответ 2.** min - 0.000000	2001.000000	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000
  max - 9016.000000	2014.000000	5076.000000	10626.000000	2469.000000	10001.000000	4970.000000	23278.000000	83.000000



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [29]:

#Вывести типы колонок тут
df.DISTRICT = df.DISTRICT.astype("category")
df.Year = df.Year.astype("int")
df.Rape = df.Rape.astype("Float64")

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10677 entries, 0 to 10676
Data columns (total 11 columns):
 #   Column                                               Non-Null Count  Dtype   
---  ------                                               --------------  -----   
 0   Unnamed: 0                                           10677 non-null  int64   
 1   STATE/UT                                             10677 non-null  object  
 2   DISTRICT                                             10677 non-null  category
 3   Year                                                 10677 non-null  int64   
 4   Rape                                                 10677 non-null  Float64 
 5   Kidnapping and Abduction                             10677 non-null  int64   
 6   Dowry Deaths                                         10677 non-null  int64   
 7   Assault on women with intent to outrage her modesty  10677 non-null  int64   
 8   Insult to modesty of Women                           106

Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

Unnamed: 0 - порядковый; Kidnapping and Abduction - количественный; STATE/UT - категориальный

> Блок с отступами



##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [32]:
#Фильтр колонки 1
df[df["Unnamed: 0"] > 9000]

Unnamed: 0     STATE/UT     DISTRICT  Year   Rape  \
9001        9001        DELHI  IGI AIRPORT  2012    0.0   
9002        9002        DELHI    NEW DELHI  2012    5.0   
9003        9003        DELHI        NORTH  2012   14.0   
9004        9004        DELHI   NORTH-EAST  2012   78.0   
9005        9005        DELHI   NORTH-WEST  2012   64.0   
9006        9006        DELHI        OUTER  2012   67.0   
9007        9007        DELHI        SOUTH  2012   70.0   
9008        9008        DELHI   SOUTH-EAST  2012  116.0   
9009        9009        DELHI   SOUTH-WEST  2012   87.0   
9010        9010        DELHI          STF  2012    0.0   
9011        9011        DELHI         WEST  2012   79.0   
9012        9012  LAKSHADWEEP  LAKSHADWEEP  2012    0.0   
9013        9013  LAKSHADWEEP        TOTAL  2012    0.0   
9014        9014   PUDUCHERRY     KARAIKAL  2012    6.0   
9015        9015   PUDUCHERRY   PUDUCHERRY  2012    7.0   
9016        9016   PUDUCHERRY        TOTAL  2012   13.0   

      Kidnapping and Abduction  Dowry Deaths  \
9001                         0             0   
9002                        13             0   
9003                        84             6   
9004                       263            25   
9005                       184            12   
9006                       317            15   
9007                       134            17   
9008                       306            11   
9009                       244            18   
9010                         0             0   
9011                       256            18   
9012                         0             0   
9013                         0             0   
9014                         2             0   
9015                        14             0   
9016                        16             0   

      Assault on women with intent to outrage her modesty  \
9001                                                  6     
9002                                                 29     
9003                                                 29     
9004                                                 69     
9005                                                 19     
9006                                                 79     
9007                                                 59     
9008                                                160     
9009                                                 69     
9010                                                  0     
9011                                                 53     
9012                                                  1     
9013                                                  1     
9014                                                  2     
9015                                                  7     
9016                                                  9     

      Insult to modesty of Women  Cruelty by Husband or his Relatives  \
9001                           0                                    0   
9002                           7                                    4   
9003                          10                                   70   
9004                           9                                  293   
9005                          14                                  212   
9006                          18                                  244   
9007                          36                                   68   
9008                          43                                  123   
9009                          21                                  207   
9010                           0                                    0   
9011                           6                                  327   
9012                           0                                    1   
9013                           0                                    1   
9014                           0                                    1   
9015                           2                    

In [33]:
#Фильтр колонки 2
df[df["STATE/UT"] == "ANDHRA PRADESH"]

Unnamed: 0        STATE/UT        DISTRICT  Year    Rape  \
0              0  ANDHRA PRADESH        ADILABAD  2001    50.0   
1              1  ANDHRA PRADESH       ANANTAPUR  2001    23.0   
2              2  ANDHRA PRADESH        CHITTOOR  2001    27.0   
3              3  ANDHRA PRADESH        CUDDAPAH  2001    20.0   
4              4  ANDHRA PRADESH   EAST GODAVARI  2001    23.0   
...          ...             ...             ...   ...     ...   
8235        8235  ANDHRA PRADESH    VIZIANAGARAM  2012    44.0   
8236        8236  ANDHRA PRADESH        WARANGAL  2012    56.0   
8237        8237  ANDHRA PRADESH  WARANGAL URBAN  2012    30.0   
8238        8238  ANDHRA PRADESH   WEST GODAVARI  2012   130.0   
8239        8239  ANDHRA PRADESH           TOTAL  2012  1341.0   

      Kidnapping and Abduction  Dowry Deaths  \
0                           30            16   
1                           30             7   
2                           34            14   
3                           20            17   
4                           26            12   
...                        ...           ...   
8235                        20             3   
8236                        58            20   
8237                        39             9   
8238                        59            19   
8239                      1403           504   

      Assault on women with intent to outrage her modesty  \
0                                                   149     
1                                                   118     
2                                                   112     
3                                                   126     
4                                                   109     
...                                                 ...     
8235                                                 96     
8236                                                117     
8237                                                 71     
8238                                                296     
8239                                               4816     

      Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                             34                                  175   
1                             24                                  154   
2                             83                                  186   
3                             38                                   57   
4                             58                                  247   
...                          ...                                  ...   
8235                         103                                  363   
8236                          34                                  101   
8237                          74                                  267   
8238                         235                                  757   
8239                        3714                                13389   

      Importation of Girls  
0                        0  
1                        0  
2                        0  
3                        0  
4                        0  
...                    ...  
8235                     0  
8236                     0  
8237                     0  
8238                     0  
8239                     0  

[370 rows x 11 columns]

In [38]:
#Фильтр колонки 3
df[df["STATE/UT"] != "WARANGAL"]

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
0               0  ANDHRA PRADESH           ADILABAD  2001  50.0   
1               1  ANDHRA PRADESH          ANANTAPUR  2001  23.0   
2               2  ANDHRA PRADESH           CHITTOOR  2001  27.0   
3               3  ANDHRA PRADESH           CUDDAPAH  2001  20.0   
4               4  ANDHRA PRADESH      EAST GODAVARI  2001  23.0   
...           ...             ...                ...   ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014   1.0   
10673         833     Lakshadweep  Total District(s)  2014   1.0   
10674         834      Puducherry           Karaikal  2014   3.0   
10675         835      Puducherry         Puducherry  2014   7.0   
10676         836      Puducherry  Total District(s)  2014  10.0   

       Kidnapping and Abduction  Dowry Deaths  \
0                            30            16   
1                            30             7   
2                            34            14   
3                            20            17   
4                            26            12   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
0                                                    149     
1                                                    118     
2                                                    112     
3                                                    126     
4                                                    109     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                              34                                  175   
1                              24                                  154   
2                              83                                  186   
3                              38                                   57   
4                              58                                  247   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[10677 rows x 11 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [42]:
#Группировка
df.groupby(["Year"]).mean()

<ipython-input-42-8dbcc4befc32>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby(["Year"]).mean()


Unnamed: 0       Rape  Kidnapping and Abduction  Dowry Deaths  \
Year                                                                  
2001       357.5  44.902235                 40.907821     19.136872   
2002      1075.0  45.543811                 40.350487     18.976356   
2003      1798.5  43.535714                 36.527473     17.054945   
2004      2527.0  50.021948                 42.737997     19.275720   
2005      3258.0  50.092769                 42.974079     18.518417   
2006      3994.5  52.291892                 47.064865     20.589189   
2007      4736.0   55.81965                 54.955585     21.784657   
2008      5488.0  56.417871                 60.286465     21.477004   
2009      6252.0  55.794003                 67.121252     21.859192   
2010      7025.0  56.924262                 76.495507     21.543004   
2011      7810.0   61.20354                 89.924147     21.790139   
2012      8611.0  61.462392                 94.357583     20.303329   
2013       411.0  81.912515                126.077764     19.642770   
2014       418.0  87.777778                136.943847     20.203106   

      Assault on women with intent to outrage her modesty  \
Year                                                        
2001                                          95.318436     
2002                                          94.417246     
2003                                          90.491758     
2004                                          94.834019     
2005                                          93.246930     
2006                                          98.964865     
2007                                         104.263795     
2008                                         106.210250     
2009                                         100.941330     
2010                                         104.269576     
2011                                         108.642225     
2012                                         111.839704     
2013                                         171.905225     
2014                                         196.499403     

      Insult to modesty of Women  Cruelty by Husband or his Relatives  \
Year                                                                    
2001                   27.223464                           137.346369   
2002                   28.247566                           136.959666   
2003                   33.859890                           139.293956   
2004                   27.437586                           159.454047   
2005                   27.241473                           159.124147   
2006                   26.935135                           170.616216   
2007                   29.475101                           204.387618   
2008                   32.099869                           213.781866   
2009                   28.706649                           233.496741   
2010                   25.573813                           241.440308   
2011                   21.668774                           250.657396   
2012                   22.621455                           262.705302   
2013                   30.592953                           288.860267   
2014                   23.261649                           293.612903   

      Importation of Girls  
Year                        
2001              0.318436  
2002              0.211405  
2003              0.126374  
2004              0.244170  
2005              0.406548  
2006              0.181081  
2007              0.164199  
2008              0.176084  
2009              0.125163  
2010              0.092426  
2011              0.202276  
2012              0.145499  
2013              0.075334  
2014              0.031063

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация